### PR TITLE
DAOS-5279 rebuild: add rebuild inflight control

### PR DESCRIPTION
**Pull request by wangdi1**
Add migration inflight control to avoid using up
all DMA.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

File: src/object/srv_obj_migrate.c [hunk 1]
```diff
@@ -1220,9 +1227,42 @@ migrate_one_ult(void *arg)
...
+		D_DEBUG(DB_REBUILD, "skipe empty mrone %p\n", mrone);
+		D_GOTO(out, rc);
+	}
+
+	D_DEBUG(DB_REBUILD, "mrone %p inflight iod %d max iod %d\n",
+		mrone, tls->mpt_inflight_iod, tls->mpt_inflight_max_iod);
+
+	while (tls->mpt_inflight_iod + mrone->mo_iod_num >=
+	       tls->mpt_inflight_max_iod && tls->mpt_inflight_max_iod != 0
+	       && !tls->mpt_fini) {
```
Comment (liuxuezhao) [inline src/object/srv_obj_migrate.c:1243]: may consider to only create migrate_one_ult when not-busy case later, as currently still create all ULTs one time that may cause ABT scheduling pressure.

**Review by liuxuezhao — APPROVED** [review]

**Review by NiuYawei — CHANGES_REQUESTED** [review]

**Comment by wangdi1** [comment]
@NiuYawei  if you have new bio export API I can use, I can update the patch. Thanks

**Review by liuxuezhao — APPROVED** [review]

File: src/object/srv_obj_migrate.c (hunk 1 above)
Comment (wangdi1) [reply src/object/srv_obj_migrate.c:1200]: yes, the new update will control ULT count as well here.

File: src/object/srv_obj_migrate.c [hunk 2]
```diff
@@ -419,7 +431,8 @@ int migrate_pool_tls_create_one(void *data)
 	pool_tls->mpt_max_eph = arg->max_eph;
 	pool_tls->mpt_pool = ds_pool_child_lookup(arg->pool_uuid);
 	pool_tls->mpt_clear_conts = arg->clear_conts;
-
+	pool_tls->mpt_inflight_max_iod = bio_get_max_chk_cnt();
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/object/srv_obj_migrate.c]: ok, I will remove these.

**Review by liuxuezhao — DISMISSED** [review]

File: src/object/srv_obj_migrate.c [hunk 3]
```diff
@@ -1176,9 +1188,42 @@ migrate_one_ult(void *arg)
 		goto out;
 	}
 
-	rc = migrate_dkey(tls, mrone);
-	D_DEBUG(DB_REBUILD, DF_UOID" migrate dkey "DF_KEY" rc %d\n",
-		DP_UOID(mrone->mo_oid), DP_KEY(&mrone->mo_dkey), rc);
+	data_size = daos_iods_len(mrone->mo_iods, mrone->mo_iod_num);
```
Comment (NiuYawei) [inline src/object/srv_obj_migrate.c:1191]: daos_iods_len() could return -1, it's not well handled in this patch.

File: src/object/srv_obj_migrate.c [hunk 4]
```diff
@@ -2298,6 +2351,27 @@ migrate_cont_iter_cb(daos_handle_t ih, d_iov_t *key_iov,
...
+		D_ASSERT(tls->mpt_obj_generated_ult >=
+			 tls->mpt_obj_executed_ult);
+		D_ASSERT(tls->mpt_generated_ult >= tls->mpt_executed_ult);
+
+		ult_cnt = min(tls->mpt_obj_generated_ult -
+			      tls->mpt_obj_executed_ult,
+			      tls->mpt_generated_ult -
+			      tls->mpt_executed_ult);
+
+		while (ult_cnt >= tls->mpt_inflight_max_ult && !tls->mpt_fini) {
```
Comment (NiuYawei) [inline src/object/srv_obj_migrate.c:2365]: How can this work? 'ult_cnt' isn't re-calculated, so the loop will never break?

File: src/object/srv_obj_migrate.c (hunk 1 above)
Comment (NiuYawei) [reply src/object/srv_obj_migrate.c:1204]: I agree with Xuehzao. The wait condition check of max size and max ULT count should be put together.

**Review by NiuYawei — CHANGES_REQUESTED** [review]

File: src/object/srv_obj_migrate.c (hunk 3 above)
Comment (wangdi1) [reply src/object/srv_obj_migrate.c:1191]: oh, it is impossible for data_size to be -1 here for rebuild. ok, I will add check here.

File: src/object/srv_obj_migrate.c (hunk 1 above)
Comment (wangdi1) [reply src/object/srv_obj_migrate.c:1204]: Hmm, we do not know the size when creating ULT to enumerate the object.

File: src/object/srv_obj_migrate.c (hunk 4 above)
Comment (wangdi1) [reply src/object/srv_obj_migrate.c:2365]: oh, right, thanks.

File: src/object/srv_obj_migrate.c (hunk 1 above)
Comment (NiuYawei) [reply src/object/srv_obj_migrate.c:1205]: but you know the per-xstream inflight ULTs here, right?

**Review by NiuYawei — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Comment by wangdi1** [comment]
rebase to resolve conflicts.

**Review by liuxuezhao — APPROVED** [review]

**Review by NiuYawei — DISMISSED** [review]

**Review by liuxuezhao — DISMISSED** [review]

**Review by NiuYawei — APPROVED** [review]

**Review by jolivier23 — DISMISSED** [review]

**Review by johannlombardi — DISMISSED** [review]

**Comment by wangdi1** [comment]
fix CI test failure.

**Review by liuxuezhao — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**